Code/Mantid/Framework/API/src/Algorithm.cpp

```diff
@@ -1324,8 +1338,24 @@ bool Algorithm::processGroups() {
               dynamic_cast<Property *>(m_pureOutputWorkspaceProps[owp])) {
         // Default name = "in1_in2_out"
         std::string outName = outputBaseName + "_" + prop->value();
+
+        Property *inputProperty = NULL;
+        WorkspacePropertyValueIs comp(prop->value());
+        auto inputProp = std::find_if(m_inputWorkspaceProps.begin(),
+                                      m_inputWorkspaceProps.end(), comp);
+        if (inputProp != m_inputWorkspaceProps.end())
+          inputProperty = dynamic_cast<Property *>(*inputProp);
+
+        // Overwrite workspaces in any input property if they have the same
+        // name as an output (i.e. copy name button in algorithm dialog used)
+        // (only need to do this for a single input, multiple will be handled
+        // by ADS)
+        if (inputProperty && inputProperty->value() == prop->value())
+          outName = m_groups[inputProp - m_inputWorkspaceProps.begin()][entry]
+                        ->name();
+
         // Except if all inputs had similar names, then the name is "out_1"
```

Comment:
@mantid-roman It protects against the ```dynamic_cast``` failing, although there shouldn't really be a case where it does.